##### Libraries ----
library(SingleCellExperiment)
library(SC3)
library(scater)
#### Input ----
# Raw reads counting matrix:
trial<- read.csv(("/home/rpg18/Desktop/SRP_article/PIPELINE/Pipeline_SecondSteps/Merged_out_mmquant.csv"), sep='\t')
# Extra article's information:
pipe<- read.csv(("/home/rpg18/Desktop/SRP_article/PIPELINE/Pipeline_SecondSteps/pipe2_SRR.tsv"), sep='\t')

# Matrix preparation for sce object creation
trial_genes <- trial$Gene
rownames(trial) <- trial_genes
trial <- subset(trial, select = -Gene)
dim(trial)

### sce_Object ----
# Matrix:
cts <- as.matrix(trial)

# sce Object creation:
sce<- SingleCellExperiment(
  assays = list(
    counts = cts),
  colData = pipe
)
# sce object
dim(sce)
# raw counting matrix
dim(cts)

## mito_Genes ----
# mito_Genes identification:
mito.genes <- grep(pattern = "^MT-", x = rownames(cts), value = TRUE)
length(mito.genes) # number of mitochondrial genes
# Percentage mito_Genes
percent.mito <- Matrix::colSums(cts[mito.genes, ])/Matrix::colSums(cts)
sce$percent.mito <- percent.mito
sce

# writes all its results obtained for cells to the colData slot of the sce object, adding additional columns
col_data <- colData(sce)
head(col_data[ , grep("sc3_", colnames(col_data))])

### QC_metrics ----
# Scater-QC:
sce <- calculateQCMetrics(sce, 
                          feature_controls = list(mito = 500:1000),
                          cell_controls = list(empty = 1:5, damaged = 31:40))
                          #, compact = TRUE)
# if compact = TRUE, graphs are not working. just delete that part for the graphs.

# writes all its results obtained for cells to the colData slot of the sce object, adding additional columns
col_data <- colData(sce)
head(col_data[ , grep("sc3_", colnames(col_data))])
sce$total_counts
# Store scater output:
qc<- sce$scater_qc

# Check sce object:
colnames(colData(sce))                         
colnames(rowData(sce))

### Normalisation ----
# Normalisation by library size: 
sizeFactors(sce) <- librarySizeFactors(sce)
summary(sizeFactors(sce))
# Scater-default:
sce <- normalize(sce, exprs_values = "counts", return_log = TRUE,
                      log_exprs_offset = NULL, centre_size_factors = TRUE,
                      preserve_zeroes = FALSE)
# Scater-CPM:
cpm(sce) <- calculateCPM(sce)

# Store log normalisation:
log<- logcounts(sce)

# Check normalisation:
assayNames(sce)
identical(exprs(sce), logcounts(sce))
counts(sce)[,1]

## Filtering_ambiguity ----
# Filter merged_genes generated by mmquant:
merged_genes<- grep(pattern = "[-]{2}", x = rownames(cts), value = TRUE) # cts is the raw matrix
length(merged_genes)
filtered <-sce[!row.names(sce)%in%merged_genes,]
rownames(filtered)

# Check filtering
dim(sce) # raw matrix
dim(filtered) # filtered



### SC3-clustering(see-Supplementary_code) ----
# Define feature names in feature_symbol column:
rowData(filtered)$feature_symbol <- rownames(filtered)
# Remove features with duplicated names:
filtered <- filtered[!duplicated(rowData(filtered)$feature_symbol), ]

# Clustering estimation:
filtered <- sc3_estimate_k(filtered) 
str(metadata(filtered)$sc3) # k=13 estimated for mmquant -> filtered

# Clustering:
filtered <- sc3(filtered, ks = 2:15, biology= TRUE, n_cores = 4) # biology = TRUE allows outliers and marker_genes identification (cell and gene filters)
dim(filtered)

# Define feature names in feature_symbol column
rowData(filtered)$feature_symbol <- rownames(filtered)
# Remove features with duplicated names
filtered <- filtered[!duplicated(rowData(filtered)$feature_symbol), ]

## SC3-solution ----
# Quantitative measure of the diagonality of the consensus matrix:
sc3_plot_silhouette(filtered, k = 13)

# Cluster stability:
sc3_plot_cluster_stability(filtered, k = 13)

# PCA_plots:
plotPCA(filtered, colour_by = "age")
plotPCA(filtered, colour_by = "sc3_2_clusters")
plotPCA(filtered, colour_by = "sc3_8_clusters")
plotPCA(filtered, colour_by = "cell_type")
plotPCA(filtered, colour_by = "sc3_9_clusters")
plotPCA(filtered, colour_by = "sc3_13_clusters")
plotPCA(filtered, colour_by = "sc3_14_clusters")

## Visualitation_Exprs ----
# Frequence of expression as a function of the mean:
plotExprsFreqVsMean(sce) # before filtering
plotExprsFreqVsMean(filtered) # after filtering

## Top_50_before_Filt ----
plotHighestExprs(sce, exprs_values = "counts")

## Top_50_after_Filt ----
plotHighestExprs(filtered, exprs_values = "counts")


## Export_results ----
# writes all its results obtained for cells to the colData slot of the sce object, adding additional columns
col_data <- colData(filtered)
head(col_data[ , grep("sc3_", colnames(col_data))])

# Export results:
sc3_export_results_xls(filtered)

#setwd("")
output_filtered <- colData(filtered)
write.table(output_filtered, file="info_mmquant_FILTERED.csv", sep='\t', col.names = NA)

## Graphs-SC3_functions ----
# PCA_outliers:
plotPCA(
  filtered, 
  colour_by = "sc3_13_clusters", 
  size_by = "sc3_13_log2_outlier_score"
)

# Plot_functions:
sc3_plot_consensus(filtered, k = 15)

sc3_plot_consensus(
  filtered, k = 13, 
  show_pdata = c(
    "cell_type", 
    "log10_total_features_by_counts",
    "sc3_13_clusters", 
    "sc3_13_log2_outlier_score"
  )
)

# Expression matrix after cell and gene filters:
sc3_plot_expression(filtered, k = 14)

sc3_plot_expression(
  filtered, k = 3, 
  show_pdata = c(
    "cell_type", 
    "log10_total_features",
    "sc3_3_clusters", 
    "sc3_3_log2_outlier_score"
  )
)

sc3_plot_expression(
  filtered, k = 9, 
  show_pdata = c(
    "cell_type", 
    "log10_total_features",
    "sc3_9_clusters", 
    "sc3_9_log2_outlier_score"
  )
)

sc3_plot_expression(
  filtered, k = 12, 
  show_pdata = c(
    "cell_type", 
    "log10_total_features",
    "sc3_12_clusters", 
    "sc3_12_log2_outlier_score"
  )
)

# DE_genes:
sc3_plot_de_genes(filtered, k = 13)

sc3_plot_de_genes(
  filtered, k = 2, 
  show_pdata = c(
    "cell_type", 
    "log10_total_features",
    "sc3_13_clusters", 
    "sc3_13_log2_outlier_score"
  )
)

sc3_plot_de_genes(
  filtered, k = 2, 
  show_pdata = c(
    "age", 
    "log10_total_features",
    "sc3_6_clusters", 
    "sc3_6_log2_outlier_score"
  )
)
# Marker_genes:
sc3_plot_markers(filtered, k = 13)

sc3_plot_markers(
  filtered, k = 6, 
  show_pdata = c(
    "cell_type", 
    "log10_total_features",
    "sc3_9_clusters", 
    "sc3_9_log2_outlier_score"
  )
)

sc3_plot_markers(
  filtered, k = 5, 
  show_pdata = c(
    "age", 
    "log10_total_features",
    "sc3_13_clusters", 
    "sc3_13_log2_outlier_score"
  )
)
# END



## Graphs-TSNE ----
# TSNE_ncomponents:
plotTSNE(filtered, ncomponents = 3, colour_by = "age")
plotTSNE(filtered, ncomponents = 3, colour_by = "cell_type")
plotTSNE(filtered, ncomponents = 3, colour_by = "sc3_2_clusters")
plotTSNE(filtered, ncomponents = 3, colour_by = "sc3_3_clusters")
plotTSNE(filtered, ncomponents = 3, colour_by = "sc3_9_clusters")
plotTSNE(filtered, ncomponents = 3, colour_by = "sc3_10_clusters")
plotTSNE(filtered, ncomponents = 3, colour_by = "sc3_13_clusters")

## More_graphs-Scater ----
# Comparing with article's data:
plotColData(filtered, x = "total_features_by_counts",
            y = "pct_counts_feature_control", colour = "cell_type") +
  theme(legend.position = "top") +
  stat_smooth(method = "lm", se = FALSE, size = 2, fullrange = TRUE)

# Scater_plotting:
plotScater(filtered, block1 = "age", block2 = "cell_type",
           colour_by = "sc3_9_clusters", nfeatures = 200, exprs_values = "counts")

# RowData:
plotRowData(filtered, x = "n_cells_by_counts_non_control", y = "mean_counts")

# Multiplot_:
p1 <- plotColData(filtered, x = "total_counts", 
                  y = "total_features_by_counts")
p2 <- plotColData(filtered, x = "pct_counts_feature_control",
                  y = "total_features_by_counts")
p3 <- plotColData(filtered, x = "pct_counts_feature_control",
                  y = "pct_counts_in_top_50_features")
multiplot(p1, p2, p3, cols = 3)

# Multiplot_:
p1 <- plotHighestExprs(filtered[, !filtered$is_cell_control])
p2 <- plotHighestExprs(filtered[, filtered$is_cell_control])
multiplot(p1, p2, cols = 2)

### Specific_gene_analysis:

## ----plot-expression-scatter-----------------------------------------------
plotExpression(filtered, rownames(filtered)[10:15],
               x = "B2M")

## ----plot-expression-col---------------------------------------------------
plotExpression(filtered, rownames(filtered)[1:6],
               colour_by = "age", 
               size_by = "B2M")

## ----plot-expression-theme-bw----------------------------------------------
plotExpression(filtered, rownames(filtered)[7:12],
               exprs_values = "counts", 
               colour = "sc3_13_clusters", show_median = TRUE, 
               log = TRUE)
  
## ----plot-expression-many--------------------------------------------------
plotExpression(filtered, rownames(filtered)[1:6])

## --------------------------------------------------------------------------
example_sce <- runPCA(filtered)
reducedDimNames(example_sce)
plotPCA(example_sce)

## ----plot-reduceddim-4comp-colby-shapeby-----------------------------------
plotReducedDim(example_sce, use_dimred = "PCA", 
               colour_by = "sc3_13_clusters") #shape_by = ""

## ----plot-reduceddim-4comp-colby-sizeby-exprs------------------------------
plotReducedDim(example_sce, use_dimred = "PCA", 
               colour_by = "B2M", size_by = "HLA-A")

## ----plot-pca-default------------------------------------------------------
plotPCA(example_sce)

## ----plot-pca-feature-controls---------------------------------------------
example_sce2 <- runPCA(filtered, 
                       feature_set = rowData(example_sce)$is_feature_control)
plotPCA(example_sce2)

## ----plot-pca-4comp-colby-shapeby------------------------------------------
example_sce <- runPCA(filtered, ncomponents=20)
plotPCA(example_sce, ncomponents = 4, colour_by = "age")
#shape_by = "cell_type")

## ----plot-pca-4comp-colby-sizeby-exprs-------------------------------------
plotPCA(example_sce, colour_by = "ARID5B", size_by = "B2M")

## ----plot-tsne-1comp-colby-sizeby-exprs------------------------------------
# Perplexity of 10 just chosen here arbitrarily. 
set.seed(1000)
example_sce <- runTSNE(example_sce, perplexity=10)
plotTSNE(example_sce, colour_by = "sc3_14_clusters") #, size_by = "A1BG ")

## ----plot-tsne-from-pca----------------------------------------------------
set.seed(1000)
example_sce <- runTSNE(example_sce, perplexity=10, use_dimred="PCA", n_dimred=10)
plotTSNE(example_sce, colour_by="sc3_14_clusters")

## ----plot-difmap-1comp-colby-sizeby-exprs----------------------------------
#BiocManager::install("destiny", lib.loc="~/R/x86_64-pc-linux-gnu-library/3.6")
library(destiny)
example_sce <- runDiffusionMap(filtered)
plotDiffusionMap(example_sce, colour_by = "sc3_13_clusters")#, size_by = "Gene_1000")


### Supplementary_code(SC3) ----
## Clustering-preparation:
# Define feature names in feature_symbol column:
rowData(filtered)$feature_symbol <- rownames(filtered)
# Remove features with duplicated names:
filtered <- filtered[!duplicated(rowData(filtered)$feature_symbol), ]

# Preparaion:
filtered <- sc3_prepare(filtered)
str(metadata(filtered)$sc3)
rowData(filtered)$sc3_gene_filter
length(rowData(filtered)$sc3_gene_filter) # Check gene_filter: FALSE or TRUE genes
dim(filtered)

## Clustering-estimation: 
filtered <- sc3_estimate_k(filtered) 
str(metadata(filtered)$sc3) # k=13 estimated for mmquant -> filtered

## Distance-calculation between the cells (distance matrices):
filtered <- sc3_calc_dists(filtered)
names(metadata(filtered)$sc3$distances)

## Transformation-calculation of the distance matrices:
filtered <- sc3_calc_transfs(filtered)
names(metadata(filtered)$sc3$transformations)
metadata(filtered)$sc3$distances
sc3_interactive(filtered)

## Clustering-step:
filtered <- sc3_kmeans(filtered, ks = 2:15) # num_cells 2,000, k = 50
names(metadata(filtered)$sc3$kmeans)

## Consensus-calculation for each value of k:
col_data <- colData(filtered)
head(col_data[ , grep("sc3_", colnames(col_data))])

filtered <- sc3_calc_consens(filtered)
names(metadata(filtered)$sc3$consensus)
names(metadata(filtered)$sc3$consensus$`14`)
metadata(filtered)$sc3$kmeans

col_data <- colData(filtered)
head(col_data[ , grep("sc3_", colnames(col_data))])

## Biology-calculation:
filtered <- sc3_calc_biology(filtered, ks = 2:14)

## Cell-outliers:
col_data <- colData(filtered)
head(col_data[ , grep("sc3_", colnames(col_data))])
#?get_outl_cells

## DE and marker_genes:
row_data <- rowData(filtered)
head(row_data[ , grep("sc3_", colnames(row_data))])
#?get_de_genes
#?get_marker_genes

## Clustering-solution:
col_data <- colData(filtered)
colData(filtered)
col_data$sc3_14_clusters
head(col_data[ , grep("sc3_", colnames(col_data))])

### Not_included(top25)----
info <- read.csv("/home/rpg18/Desktop/SRP_article/PIPELINE/Pipeline_SecondSteps/Sc3/Results_attempts_trial/Authors/NEW/new.R/Norm_postfilter/info_mmquant_FILTERED.csv", sep='\t')

sample <- info$Run
rownames(info) <- sample

info <- subset(info, select = -Run)
cluster<- info$sc3_13_clusters

lista<- cbind(sample,cluster)

lista<- DataFrame(lista)
lista$sample
rownames(lista)<- sample
lista<- subset(lista, -sample)
lista
cluster1 <-c()
cluster2 <-c()
cluster3 <-c()
cluster4 <-c()
cluster5 <-c()
cluster6 <-c()
cluster7 <-c()
cluster8 <-c()
cluster9 <-c()

row.names(lista)

count = 1
for(k in row.names(lista)){
  meh <- info$sc3_9_clusters[count]
  count <- count + 1
  if (meh %in% 1){cluster1<-append(cluster1,k)}
  else if (meh %in% 2){cluster2<-append(cluster2,k)}
  else if (meh %in% 3){cluster3<-append(cluster3,k)}
  else if (meh %in% 4){cluster4<-append(cluster4,k)}
  else if (meh %in% 5){cluster5<-append(cluster5,k)}
  else if (meh %in% 6){cluster6<-append(cluster6,k)}
  else if (meh %in% 7){cluster7<-append(cluster7,k)}
  else if (meh %in% 8){cluster8<-append(cluster8,k)}
  else if (meh %in% 9){cluster9<-append(cluster9,k)}
  
}

outputsm<- cts

hopeful1<-outputsm[,cluster1]
hopeful2<-outputsm[,cluster2]
hopeful3<-outputsm[,cluster3]
hopeful4<-outputsm[,cluster4]
hopeful5<-outputsm[,cluster5]
hopeful6<-outputsm[,cluster6]
hopeful7<-outputsm[,cluster7]
hopeful8<-outputsm[,cluster8]
hopeful9<-outputsm[,cluster9]

total<- rowSums(hopeful1)
hopeful1<-cbind(hopeful1,total)
hopeful1
ordered1 <- hopeful1[order(-total),]
genes1 <-row.names(ordered1[1:25,])
ordered1

total<- rowSums(hopeful2)
total
hopeful2<-cbind(hopeful2,total)
ordered2 <- hopeful2[order(-total),]
genes2 <-row.names(ordered2[1:25,])

total<- rowSums(hopeful3)
hopeful3<-cbind(hopeful3,total)
ordered3 <- hopeful3[order(-total),]
genes3 <-row.names(ordered3[1:25,])

total<- rowSums(hopeful4)
hopeful4<-cbind(hopeful4,total)
ordered4 <- hopeful4[order(-total),]
genes4 <-row.names(ordered4[1:25,])

total<- rowSums(hopeful4)
hopeful5<-cbind(hopeful5,total)
ordered5 <- hopeful5[order(-total),]
genes5 <-row.names(ordered5[1:25,])

total<- rowSums(hopeful6)
hopeful6<-cbind(hopeful6,total)
ordered6 <- hopeful6[order(-total),]
genes6 <-row.names(ordered6[1:25,])

total<- rowSums(hopeful7)
hopeful7<-cbind(hopeful7,total)
ordered7 <- hopeful7[order(-total),]
genes7 <-row.names(ordered7[1:25,])

total<- rowSums(hopeful8)
hopeful8<-cbind(hopeful8,total)
ordered8 <- hopeful8[order(-total),]
genes8 <-row.names(ordered8[1:25,])

total<- rowSums(hopeful9)
hopeful9<-cbind(hopeful9,total)
ordered9 <- hopeful9[order(-total),]
genes9 <-row.names(ordered9[1:25,])

all_clusters<- cbind(genes1,genes2,genes3,genes4,genes5,genes6,genes7,genes8,genes9)

write.table(cluster1, file = "cells_cluster1.txt")
write.table(cluster2, file = "cells_cluster2.txt")
write.table(cluster3, file = "cells_cluster3.txt")
write.table(cluster4, file = "cells_cluster4.txt")
write.table(cluster5, file = "cells_cluster5.txt")
write.table(cluster6, file = "cells_cluster6.txt")
write.table(cluster7, file = "cells_cluster7.txt")
write.table(cluster8, file = "cells_cluster8.txt")
write.table(cluster9, file = "cells_cluster9.txt")

write.table(all_clusters, file = "top25_our_genes_9_clusters.csv", col.names = NA)

### (Not_included) #####
# Just in case there is a row with all 0:
keep_feature <- rowSums(counts(sce) > 0) > 0
sce <-sce[keep_feature,]
dim(sce)

keep_feature <- nexprs(sce, byrow=TRUE) >= 2 # At least 2 cells per cluster 
sce <- sce[keep_feature,]
dim(sce)


example_sce <- runColDataPCA(filtered,
                      detect_outliers = TRUE)
plotPCA(example_sce)


## Filtering_sequencing_depth
# by counts depth:
keep.total <- sce$total_counts > 50000 
filtered <- sce[,keep.total] # filtered cell no.71 in colnames(sce): SRR1974613

# Check filtering:
dim(sce) # raw matrix
dim(filtered) # merged_genes filter



